File: skills/langchain-typescript/assets/langgraph-agent/graph-agent.ts
import { StateGraph, MessagesAnnotation, START, END } from "@langchain/langgraph";
import { ChatOpenAI } from "@langchain/openai";
import { ToolNode } from "@langchain/langgraph/prebuilt";
import { DuckDuckGoSearch } from "@langchain/community/tools/duckduckgo_search";
import { MemorySaver } from "@langchain/langgraph";

/**
 * LangGraph Agent Template
 * 
 * This template demonstrates a graph-based agent with:
 * - State management
 * - Conditional routing
 * - Tool execution
 * - Persistence
 */

// Configure tools
const tools = [
  new DuckDuckGoSearch({ maxResults: 5 })
];

const toolNode = new ToolNode(tools);

// Configure model with tool binding
const model = new ChatOpenAI({
  model: "gpt-4",
  temperature: 0
}).bindTools(tools);

// Define agent node
async function callModel(state: typeof MessagesAnnotation.State) {
  const response = await model.invoke(state.messages);
  return { messages: [response] };
}

// Define routing logic
function shouldContinue(state: typeof MessagesAnnotation.State) {
  const lastMessage = state.messages[state.messages.length - 1];
  
  // If the LLM called tools, execute them
  if ("tool_calls" in lastMessage && lastMessage.tool_calls?.length > 0) {
    return "tools";
  }
  
  // Otherwise, end the conversation
  return END;
}

// Build the graph
const workflow = new StateGraph(MessagesAnnotation)
  .addNode("agent", callModel)
  .addNode("tools", toolNode)
  .addEdge(START, "agent")
  .addConditionalEdges("agent", shouldContinue, {
    tools: "tools",
    [END]: END
  })
  .addEdge("tools", "agent");  // Loop back after tool execution

// Compile with persistence
const checkpointer = new MemorySaver();
const graph = workflow.compile({ checkpointer });

// Run the agent
async function main() {
  const userQuery = "What are the latest developments in AI?";
  
  console.log(`User: ${userQuery}\n`);
  
  // Create a thread for this conversation
  const config = {
    configurable: {
      thread_id: "conversation-1"
    }
  };
  
  // Invoke with streaming
  const stream = await graph.stream(
    {
      messages: [{ role: "user", content: userQuery }]
    },
    {
      ...config,
      streamMode: "values"
    }
  );
  
  // Process stream
  for await (const event of stream) {
    const lastMessage = event.messages[event.messages.length - 1];
    if (lastMessage.content) {
      console.log(`\n${lastMessage._getType()}: ${lastMessage.content}`);
    }
  }
  
  // Continue conversation
  console.log("\n--- Second turn ---\n");
  
  const followUp = await graph.invoke(
    {
      messages: [{ role: "user", content: "Can you summarize that?" }]
    },
    config  // Same thread maintains context
  );
  
  const finalMessage = followUp.messages[followUp.messages.length - 1];
  console.log(`\nAssistant: ${finalMessage.content}`);
}

// Run if executed directly
if (require.main === module) {
  main().catch(console.error);
}

export { graph };

File: skills/langchain-python/assets/langgraph-agent/graph_agent.py
"""
LangGraph Agent with State Management

An advanced agent using LangGraph for complex workflows with:
- State management
- Conditional routing
- Persistence
- Human-in-the-loop support
"""

import os
from dotenv import load_dotenv
from typing import TypedDict, Annotated
from operator import add

from langgraph.graph import StateGraph, START, END
from langgraph.checkpoint.memory import MemorySaver
from langgraph.prebuilt import create_react_agent
from langchain_openai import ChatOpenAI
from langchain.tools import tool

# Load environment variables
load_dotenv()

# Define state schema
class AgentState(TypedDict):
    """State for the agent workflow."""
    messages: Annotated[list, add]  # Accumulates messages
    iterations: int
    result: str
    status: str

# Define tools
@tool
def research_tool(query: str) -> str:
    """Research information on a topic."""
    return f"Research results for: {query}"

@tool
def analysis_tool(data: str) -> str:
    """Analyze data and extract insights."""
    return f"Analysis of: {data}"

@tool
def calculator(expression: str) -> str:
    """Calculate mathematical expressions."""
    try:
        return str(eval(expression))
    except Exception as e:
        return f"Error: {e}"

# Initialize components
llm = ChatOpenAI(model="gpt-4", temperature=0)
tools = [research_tool, analysis_tool, calculator]

# Create React agent
react_agent = create_react_agent(
    llm,
    tools,
    messages_modifier="You are a helpful assistant. Use tools when needed."
)

# Define nodes
def agent_node(state: AgentState):
    """Main agent processing node."""
    print(f"\n[Agent Node] Iteration {state['iterations'] + 1}")
    
    # Run agent
    result = react_agent.invoke(state)
    
    return {
        "messages": result["messages"],
        "iterations": state["iterations"] + 1,
        "status": "processing"
    }

def should_continue(state: AgentState) -> str:
    """Decide whether to continue or end."""
    
    # Check if we have a final answer
    last_message = state["messages"][-1]
    
    if hasattr(last_message, "tool_calls") and last_message.tool_calls:
        return "continue"
    
    # Check iteration limit
    if state["iterations"] >= 5:
        return "end"
    
    # Check if task is complete
    if "final" in last_message.content.lower():
        return "end"
    
    return "continue"

def finalize_node(state: AgentState):
    """Finalize the result."""
    print("\n[Finalize Node] Preparing final output")
    
    last_message = state["messages"][-1]
    
    return {
        "result": last_message.content,
        "status": "complete"
    }

# Build the graph
def create_workflow():
    """Create the LangGraph workflow."""
    
    workflow = StateGraph(AgentState)
    
    # Add nodes
    workflow.add_node("agent", agent_node)
    workflow.add_node("finalize", finalize_node)
    
    # Add edges
    workflow.add_edge(START, "agent")
    
    # Conditional routing
    workflow.add_conditional_edges(
        "agent",
        should_continue,
        {
            "continue": "agent",  # Loop back
            "end": "finalize"
        }
    )
    
    workflow.add_edge("finalize", END)
    
    # Compile with checkpointer for persistence
    checkpointer = MemorySaver()
    app = workflow.compile(checkpointer=checkpointer)
    
    return app

def run_agent(query: str):
    """Run the agent with a query."""
    
    print(f"\n{'='*70}")
    print(f"Query: {query}")
    print(f"{'='*70}")
    
    # Create app
    app = create_workflow()
    
    # Configure with thread ID for persistence
    config = {"configurable": {"thread_id": "example-1"}}
    
    # Initial state
    initial_state = {
        "messages": [("user", query)],
        "iterations": 0,
        "result": "",
        "status": "started"
    }
    
    # Run workflow
    final_state = app.invoke(initial_state, config)
    
    print(f"\n{'='*70}")
    print(f"Result: {final_state['result']}")
    print(f"Iterations: {final_state['iterations']}")
    print(f"Status: {final_state['status']}")
    print(f"{'='*70}\n")
    
    return final_state

def main():
    """Main function."""
    
    # Example queries
    queries = [
        "Research AI trends and calculate the market size",
        "Analyze the benefits of machine learning in healthcare",
        "What is 25 * 17 + 50?",
    ]
    
    for query in queries:
        run_agent(query)

if __name__ == "__main__":
    main()
